### PR TITLE
Change idle_timeout to max_idle_timeout

### LINKED ISSUE #2602: Idle timeout needs more description and a recommendation
As it stands, the idle timeout period is used by both QUIC endpoints, and may be made available to the application but,
- it's not clear how an endpoint should determine this value,
- it's not explicitly stated that the min(server, client) should be used,
- it's not explicitly stated that this min value should be communicated to the application, which needs it to determine how often to send PINGs.

A straw proposal is to use 30 seconds as the default idle timeout period, and state the other two points explicitly in the transport and http drafts.

### PR DESCRIPTION
**Pull request by ianswett**
This makes idle timeout symmetric by stating that a peer will close the connection at or soon after the idle timeout expires, creating clear expectations of idle timeout behavior.  This also renames the transport parameter idle_timeout to max_idle_timeout for clarity.

Fixes #2602 
Takes some of the improved text from #2745

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2357,15 +2357,22 @@ containing frames other than ACK or PADDING (an ack-eliciting packet; see
...
-that it is likely that only applications or application protocols will
-know what information can be retried.
+Each endpoint can advertise a different idle timeout value. An idle timeout
+indicates the connection will be abandoned when the timer ends or soon
+afterwards.  If a peer advertises a smaller idle timeout value than the
+local endpoint, the local endpoint MAY choose to idle timeout after the
+shorter period advertised by the peer.  By announcing an idle timeout,
+an endpoint commits to initiating an immediate close ({{immediate-close}})
+if it abandons a connection prior to both its peers and its own idle timeout
+expiring.
```
[the remark this answers is not in the document]
Comment (ianswett) [reply draft-ietf-quic-transport.md]: It's committing to explicitly closing(and not silently closing) if it's before min(local_timeout, remote_timeout).  This was adapted from MT's PR.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2357,15 +2357,22 @@ containing frames other than ACK or PADDING (an ack-eliciting packet; see
...
-endpoint.  An endpoint that sends packets near the end of the idle timeout
-period of a peer risks having those packets discarded if its peer enters the
-draining state before the packets arrive.  If a peer could timeout within a
-Probe Timeout (PTO; see Section 6.3 of {{QUIC-RECOVERY}}), it is advisable to
-test for liveness before sending any data that cannot be retried safely.  Note
-that it is likely that only applications or application protocols will
-know what information can be retried.
+Each endpoint can advertise a different idle timeout value. An idle timeout
+indicates the connection will be abandoned when the timer ends or soon
+afterwards.  If a peer advertises a smaller idle timeout value than the
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Each endpoint advertises a different idle timeout, but the effective value is determined by taking the minimum of the two advertised values.  By announcing an idle timeout, endpoints commit to initiating an immediate close ({{immediate-close}}) if it abandons a connection prior to the effective value.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -4414,11 +4416,11 @@ original_connection_id (0x0000):
...
 
-idle_timeout (0x0001):
+max_idle_timeout (0x0001):
 
-: The idle timeout is a value in milliseconds that is encoded as an integer; see
-  ({{idle-timeout}}).  If this parameter is absent or zero then the idle
-  timeout is disabled.
+: The max idle timeout is a value in milliseconds that is encoded as an integer;
+  see ({{idle-timeout}}).  If this parameter is absent or zero then the idle
+  timeout is disabled unless the peer specifies a max idle timeout.
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: Maybe rephrase to: "Idle timeout is disabled when this parameter is absent or zero on both sides"?

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -4593,11 +4595,12 @@ endpoints send PING frames without coordination can produce an excessive number
 of packets and poor performance.
 
 A connection will time out if no packets are sent or received for a period
-longer than the time specified in the idle_timeout transport parameter (see
-{{termination}}).  However, state in middleboxes might time out earlier than
-that.  Though REQ-5 in {{?RFC4787}} recommends a 2 minute timeout interval,
-experience shows that sending packets every 15 to 30 seconds is necessary to
-prevent the majority of middleboxes from losing state for UDP flows.
+longer than the time specified in the max_idle_timeout transport parameter
```
Comment (kazuho) [inline draft-ietf-quic-transport.md]: s/specified in/negotiated by using/?

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: Thanks!

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
-The value for an idle timeout can be asymmetric.  The value advertised by an
-endpoint is only used to determine whether the connection is live at that
-endpoint.  An endpoint that sends packets near the end of the idle timeout
-period of a peer risks having those packets discarded if its peer enters the
-draining state before the packets arrive.  If a peer could timeout within a
-Probe Timeout (PTO; see Section 6.3 of {{QUIC-RECOVERY}}), it is advisable to
-test for liveness before sending any data that cannot be retried safely.  Note
-that it is likely that only applications or application protocols will
-know what information can be retried.
+Each endpoint advertises a different max idle timeout, but the effective value
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
Each endpoint advertises a different max_idle_timeout, but the effective value
```

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+is determined by taking the minimum of the two advertised values. By announcing
+a max idle timeout, endpoints commit to initiating an immediate close
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
a max_idle_timeout, endpoints commit to initiating an immediate close
```

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+({{immediate-close}}) if it abandons a connection prior to the effective value.
+
+An endpoint restarts any timer it maintains when a packet from its peer is
+received and processed successfully.  The timer is also restarted when sending
+a packet containing frames other than ACK or PADDING (an ack-eliciting packet;
+see {{QUIC-RECOVERY}}), but only if no other ack-eliciting packets have been
+sent since last receiving a packet.  Restarting when sending packets ensures
+that connections do not prematurely time out when initiating new activity.
+
+An endpoint that sends packets near the end of its peer's idle timeout period
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
An endpoint that sends packets near the end of the idle timeout period
```

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -2344,28 +2344,30 @@ source address.
 
 ## Idle Timeout {#idle-timeout}
 
-If the idle timeout is enabled, a connection is silently closed and the state is
-discarded when it remains idle for longer than both the advertised
+If the idle timeout is enabled by either peer, a connection is silently closed
+and the state is discarded when it remains idle for longer than the max
 idle timeout (see {{transport-parameter-definitions}}) and three times the
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
max_idle_timeout (see {{transport-parameter-definitions}}) and three times the
```

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -2344,28 +2344,30 @@ source address.
 
 ## Idle Timeout {#idle-timeout}
 
-If the idle timeout is enabled, a connection is silently closed and the state is
-discarded when it remains idle for longer than both the advertised
+If the idle timeout is enabled by either peer, a connection is silently closed
+and the state is discarded when it remains idle for longer than the max
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
and the state is discarded when it remains idle for longer than the
```

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+is determined by taking the minimum of the two advertised values. By announcing
+a max idle timeout, endpoints commit to initiating an immediate close
+({{immediate-close}}) if it abandons a connection prior to the effective value.
+
+An endpoint restarts any timer it maintains when a packet from its peer is
+received and processed successfully.  The timer is also restarted when sending
+a packet containing frames other than ACK or PADDING (an ack-eliciting packet;
+see {{QUIC-RECOVERY}}), but only if no other ack-eliciting packets have been
+sent since last receiving a packet.  Restarting when sending packets ensures
+that connections do not prematurely time out when initiating new activity.
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: If you're going to explain why you restart when sending, you might need to explain why you _don't_ restart when sending subsequent packets.  Perhaps "when sending _one or more_ packets" might help point this explanation at bursts, rather than at individual packets?

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
-endpoint is only used to determine whether the connection is live at that
-endpoint.  An endpoint that sends packets near the end of the idle timeout
-period of a peer risks having those packets discarded if its peer enters the
-draining state before the packets arrive.  If a peer could timeout within a
-Probe Timeout (PTO; see Section 6.3 of {{QUIC-RECOVERY}}), it is advisable to
-test for liveness before sending any data that cannot be retried safely.  Note
-that it is likely that only applications or application protocols will
-know what information can be retried.
+Each endpoint advertises a different max_idle_timeout, but the effective value
+is determined by taking the minimum of the two advertised values. By announcing
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
at an endpoint is computed as the minimum of the two advertised values. By announcing
```

File: draft-ietf-quic-transport.md [hunk 12]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+a max_idle_timeout, endpoints commit to initiating an immediate close
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
a max_idle_timeout, an endpoint commits to initiating an immediate close
```

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
-The value for an idle timeout can be asymmetric.  The value advertised by an
-endpoint is only used to determine whether the connection is live at that
-endpoint.  An endpoint that sends packets near the end of the idle timeout
-period of a peer risks having those packets discarded if its peer enters the
-draining state before the packets arrive.  If a peer could timeout within a
-Probe Timeout (PTO; see Section 6.3 of {{QUIC-RECOVERY}}), it is advisable to
-test for liveness before sending any data that cannot be retried safely.  Note
-that it is likely that only applications or application protocols will
-know what information can be retried.
+Each endpoint advertises a different max_idle_timeout, but the effective value
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
Each endpoint advertises a max_idle_timeout, but the effective value
```

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+is determined by taking the minimum of the two advertised values. By announcing
+a max_idle_timeout, endpoints commit to initiating an immediate close
+({{immediate-close}}) if it abandons a connection prior to the effective value.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
({{immediate-close}}) if it abandons the connection prior to the effective value.
```

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I don't understand this sentence... what does it mean for an endpoint to initiate an immediate close if it abandons the connection prior to the min? I thought that the endpoint was to initiate an immediate close _at_ the min. Why are we talking about abandoning?

File: draft-ietf-quic-transport.md [hunk 15]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+
+An endpoint restarts any timer it maintains when a packet from its peer is
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Do you mean an idle timeout timer? I would say that explicitly. Maybe even give it a name - ITO?

File: draft-ietf-quic-transport.md [hunk 16]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+received and processed successfully.  The timer is also restarted when sending
+a packet containing frames other than ACK or PADDING (an ack-eliciting packet;
+see {{QUIC-RECOVERY}}), but only if no other ack-eliciting packets have been
+sent since last receiving a packet.  Restarting when sending packets ensures
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Can you write some rationale for these restarting rules?

File: draft-ietf-quic-transport.md [hunk 17]
```diff
@@ -2344,28 +2344,30 @@ source address.
...
+that connections do not prematurely time out when initiating new activity.
+
+An endpoint that sends packets near the end of the idle timeout period
+risks having those packets discarded if its peer enters the draining state
+before the packets arrive.  If a peer could time out within an Probe Timeout
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: This is not a great construction: "If a peer could time out within a Probe Timeout".  It's also confusing ... maybe change this to "When an endpoint is about to send data that cannot be retried safely, it is possible that the peer may have reached the end of its idle timeout period before this data arrives at the peer. To avoid such situations, it is recommended that an endpoint test for liveness when it expects that the peer might be close, within a Probe Timeout (PTO; see Section ...) period for instance, to the end of its idle timeout period."

File: draft-ietf-quic-transport.md [hunk 18]
```diff
@@ -4414,11 +4416,11 @@ original_connection_id (0x0000):
...
 
-idle_timeout (0x0001):
+max_idle_timeout (0x0001):
 
-: The idle timeout is a value in milliseconds that is encoded as an integer; see
-  ({{idle-timeout}}).  If this parameter is absent or zero then the idle
-  timeout is disabled.
+: The max idle timeout is a value in milliseconds that is encoded as an integer;
+  see ({{idle-timeout}}).  Idle timeout is disabled when this parameter is
+  absent or zero on both sides.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Rephrase -- not both sides, but when the parameter is advertized as zero by both endpoints.

File: draft-ietf-quic-transport.md (hunk 15 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: The text from here down is pre-existing, so I'd prefer not to change it as part of this PR.

File: draft-ietf-quic-transport.md (hunk 16 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: Also pre-existing text, so maybe in a different PR?

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: Also pre-existing.

File: draft-ietf-quic-transport.md (hunk 17 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: Agreed, I'll file an issue to address your and Mike's comments.

**Comment by ianswett** [comment]
@MikeBishop and @janaiyengar I filed a new issue(#3184) for your comments on the existing idle timeout text.  I hope all of those comments can be addressed in an editorial PR.

File: draft-ietf-quic-transport.md [hunk 19]
```diff
@@ -2343,28 +2343,30 @@ source address.
...
+and its state is discarded when it remains idle for longer than the
+max_idle_timeout (see {{transport-parameter-definitions}}) and three times the
 current Probe Timeout (PTO).
 
-Each endpoint advertises its own idle timeout to its peer.  An endpoint
-restarts any timer it maintains when a packet from its peer is received and
-processed successfully.  The timer is also restarted when sending an
+Each endpoint advertises a max_idle_timeout, but the effective value
+at an endpoint is computed as the minimum of the two advertised values. By announcing
+a max_idle_timeout, an endpoint commits to initiating an immediate close
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: > a max_idle_timeout, an endpoint declares that after being idle for longer than the advertised time it will not send any packets and it could generate a stateless reset ({{stateless-reset}}) in reaction to any received packet.  An endpoint commits to initiating an immediate close [...]

File: draft-ietf-quic-transport.md (hunk 15 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md]: I don't think that we need new names.  New names are best when you need to refer to them in many places, but idle timeout logic is all constrained to this one section.

File: draft-ietf-quic-transport.md [hunk 20]
```diff
@@ -2343,28 +2343,30 @@ source address.
 
 ## Idle Timeout {#idle-timeout}
 
-If the idle timeout is enabled, a connection is silently closed and the state is
-discarded when it remains idle for longer than both the advertised
-idle timeout (see {{transport-parameter-definitions}}) and three times the
+If the idle timeout is enabled by either peer, a connection is silently closed
+and its state is discarded when it remains idle for longer than the
+max_idle_timeout (see {{transport-parameter-definitions}}) and three times the
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md]: This sentence seems to contradict the next paragraph: The connection is closed when it remains idle for longer than the minimum of the max_idle_timeouts.

File: draft-ietf-quic-transport.md [hunk 21]
```diff
@@ -4439,11 +4441,11 @@ original_connection_id (0x0000):
...
 
-idle_timeout (0x0001):
+max_idle_timeout (0x0001):
 
-: The idle timeout is a value in milliseconds that is encoded as an integer; see
-  ({{idle-timeout}}).  If this parameter is absent or zero then the idle
-  timeout is disabled.
+: The max idle timeout is a value in milliseconds that is encoded as an integer;
+  see ({{idle-timeout}}).  Idle timeout is disabled when both endpoints omit this
+  transport parameteter or specify a value of 0.
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md]: In its current wording, this would mean that if one of the endpoints omits the parameter or sets it to 0, the resulting idle timeout would be 0 (since we're taking the minimum of both values).

Comment (martinthomson) [reply draft-ietf-quic-transport.md]: I think that we should not interpret 0→∞.  I would prefer that we omit a value when there is none and assign the value an infinite value (or 2^62-1 if you prefer) as a default.

File: draft-ietf-quic-transport.md (hunk 15 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: It's a new timeout. It's not covered by any existing timeouts, so calling it out explicitly avoids use of generic language below.

File: draft-ietf-quic-transport.md (hunk 17 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I think the text should be fixed here... why introduce text here and then fix it in a new PR?

File: draft-ietf-quic-transport.md (hunk 21 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: SG, I'm going to no longer make the default 0.

File: draft-ietf-quic-transport.md (hunk 17 above)
Comment (ianswett) [reply draft-ietf-quic-transport.md]: I'm not introducing this text, it just moved, hence the different PR, which is editorial.

**Comment by ianswett** [comment]
This is failing due to a QPACK error, but I believe the PR is done.

As noted above, I opened #3184 for the many comments about the existing text.  I can take a stab at this, but I'd like to do that separately from this design change.

File: draft-ietf-quic-transport.md (hunk 15 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md]:
```suggestion
An endpoint restarts its idle timer when a packet from its peer is
```

File: draft-ietf-quic-transport.md [hunk 22]
```diff
@@ -2343,28 +2343,31 @@ source address.
...
+the current Probe Timeout (PTO).
+
+Each endpoint advertises a max_idle_timeout, but the effective value
+at an endpoint is computed as the minimum of the two advertised values. By
+announcing a max_idle_timeout, an endpoint commits to initiating an immediate
+close ({{immediate-close}}) if it abandons the connection prior to the effective
+value.
+
+An endpoint restarts any timer it maintains when a packet from its peer is
+received and processed successfully.  The timer is also restarted when sending
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
received and processed successfully.  The idle timer is also restarted when sending
```

File: draft-ietf-quic-transport.md [hunk 23]
```diff
@@ -2343,28 +2343,31 @@ source address.
...
+an ack-eliciting packet (see {{QUIC-RECOVERY}}), but only if no other
+ack-eliciting packets have been sent since last receiving a packet.  Restarting
+when sending packets ensures that connections do not prematurely time out when
+initiating new activity.
+
+An endpoint that sends packets near the end of the idle timeout period
+risks having those packets discarded if its peer enters the draining state
+before the packets arrive.  If a peer could time out within a Probe Timeout
+(PTO; see Section 6.2.2 of {{QUIC-RECOVERY}}), it is advisable to test for
+liveness before sending any data that cannot be retried safely.  Note that it
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: What does it mean to test for liveness? Can you make this more explicit?
